gobblin-core/src/main/java/gobblin/publisher/BaseDataPublisher.java

```diff
@@ -336,12 +341,50 @@ private ParallelRunner getParallelRunner(FileSystem fs) {
 
   @Override
   public void publishMetadata(Collection<? extends WorkUnitState> states) throws IOException {
-    // Nothing to do
+        for (WorkUnitState workUnitState : states) {
+        publishMetadata(workUnitState);
+    }
   }
 
   @Override
   public void publishMetadata(WorkUnitState state) throws IOException {
-    // Nothing to do
-  }
+    String metadataValue = state.getProp(ConfigurationKeys.DATA_PUBLISHER_METADATA_STR);
+    if (metadataValue == null) {
+      //Nothing to write
+      return;
+    }
+
+    if (state.getProp(ConfigurationKeys.DATA_PUBLISHER_METADATA_OUTPUT_DIR) == null) {
+      LOG.error("Missing metadata output directory path : "  + ConfigurationKeys.DATA_PUBLISHER_METADATA_OUTPUT_DIR 
+                + " in the config");
+      return;
+    }
+
+    //Write for each branch
+    for (int branchId = 0; branchId < this.numBranches; branchId++) {
+      FileSystem fs = this.metaDataWriterFileSystemByBranches.get(branchId);
 
+      String fileName = ForkOperatorUtils.getPropertyNameForBranch("metadata.txt", this.numBranches, branchId);
```

Comment:
Please use a named constant and add the "data.publisher." prefix. "metadata.txt" suggest a plain-text format. I suggest choosing a name along the lines of "data.publisher.metadata.output_file".